### PR TITLE
(PE-35637) add explicit read/write locking to serial file use

### PR DESCRIPTION
**Pull request by jonathannewman**
In preparation for future changes coming to the CA endpoints, this adds explicit read/write locking to the management of the serial file.

This also adds the ability for the serial lock acquisition to timeout preventing potential locking issues. The default timeout is 5 seconds.

The locks are added to the ca-settings, and schemas updated as appropriate.

### CODE REVIEW

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj [hunk 1]
```diff
@@ -1,5 +1,6 @@
 (ns puppetlabs.puppetserver.certificate-authority
-  (:import [org.apache.commons.io IOUtils]
+  (:import (java.util.concurrent.locks ReentrantReadWriteLock)
```
Comment (jonathannewman) [inline src/clj/puppetlabs/puppetserver/certificate_authority.clj:2]: The style of `[]` vs `()` is inconsistent here.  Cursive automatically uses `()`.  I can make this consistent.  Thoughts?

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj [hunk 2]
```diff
@@ -928,14 +948,14 @@
     (fs/exists? hostpubkey)
     (throw
      (IllegalStateException.
-      (i18n/trs "Found master public key ''{0}'' but master private key ''{1}'' is missing"
+      ^String (i18n/trs "Found master public key ''{0}'' but master private key ''{1}'' is missing"
```
Comment (jonathannewman) [inline src/clj/puppetlabs/puppetserver/certificate_authority.clj:951]: Editor was complaining that this was ambiguous, so I specified the type explicitly.

**Review by steveax — APPROVED** [review]
👍

**Comment by shaigy** [comment]
👍

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj (hunk 1 above)
Comment (justinstoller) [reply src/clj/puppetlabs/puppetserver/certificate_authority.clj:2]: `()` is more idiomatic, and I'd be into making them all lists. IIRC, this code was written prior to strong standards in the clj community.

**Comment by jonathannewman** [comment]
Commit comment got messed up.  I'll fix that too.

**Review by justinstoller — APPROVED** [review]
Looks good.

**Pull request merged**